bleach/utils.py

```diff
@@ -40,3 +43,52 @@ def force_unicode(text):
 
     # If not, convert it
     return six.text_type(text, "utf-8", "strict")
+
+
+netloc_port_re = re.compile(
+    "^" + URLValidator.netloc_re + URLValidator.port_re + "$", re.IGNORECASE
+)
+
+
+# Characters valid in scheme names
+scheme_chars = (
+    "abcdefghijklmnopqrstuvwxyz" "ABCDEFGHIJKLMNOPQRSTUVWXYZ" "0123456789" "+-."
+)
+
+
+def _is_valid_netloc_and_port(netloc):
+    """
+    Returns the scheme for a URI or None when parsing the URI fails
+
+    :arg str/unicode netloc:
+
+    :returns: bool
+
+    """
+    # The maximum length of a full host name is 253 characters per RFC 1034
+    # section 3.1. It's defined to be 255 bytes or less, but this includes
+    # one byte for the length of the name and one byte for the trailing dot
+    # that's used to indicate absolute names in DNS.
+    netloc = netloc_port_re.match(netloc)
+    return bool(netloc and len(netloc.group(0)) < 254)
+
+
+def _parse_uri_scheme(uri):
+    """
+    Returns the scheme for a URI or None when parsing the URI fails
+
+    :arg str/unicode text:
+
+    :returns: text or None
+
+    """
+    # replicate Python 3.9 urlparse scheme parsing for older Python versions
+    i = uri.find(":")
+    if i > 0:
+        scheme = uri[:i]
+        for c in uri[:i]:
+            if c not in scheme_chars:
+                break
+        return scheme
+
+    return None
```

Comment:
I think I'd add tests for this, too.

Reply:
Something like:

```
@pytest.mark.parametrize("text, expected", [
    ("", None),
    ("noscheme", None),
    ("https://example.com", "https"),
    ("git+ssh://example.com", "git+ssh"),
    # FIXME: example for . and -
])
def test_parse_uri_scheme(data, expected):
    assert _parse_uri_scheme(data) == expected
```

Reply:
Ended up not trying to pull in the urlparse 3.9 changes.